repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -114,6 +116,21 @@ private void processCreateProduct(AtlasEntity entity,AtlasVertex vertex) throws
     private void processUpdateProduct(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
         AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processUpdateProduct");
 
+        String state = vertex.getProperty(STATE_PROPERTY_KEY, String.class);
+
+        if (DELETED.name().equals(state)) {
+            AtlasEntity.Status entityStatus = entity.getStatus();
+            Object entityState =  entity.getAttribute(STATE_PROPERTY_KEY);
+
+            //  To allow product restoration but block all other updates if the product is archived
+            boolean isBeingUnarchived = (entityStatus != null && AtlasEntity.Status.ACTIVE.equals(entityStatus)) ||
+                    (entityState != null && ACTIVE.name().equals(entityState.toString()));
+
+            if (!isBeingUnarchived) {
+                throw new AtlasBaseException(OPERATION_NOT_SUPPORTED, "Cannot update DataProduct that is Archived!");
+            }
```

Comment:
Loginc can be replaced by look up in context.getEntitiesToRestore() (on vertex.getIdForDisplay())

Since `context.addEntityToRestore(vertex);` is aslready called until we process Pre-processors, it should be present in context.getEntitiesToRestore()